sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/api/RestInternalUsersApiAction.java

```diff
@@ -0,0 +1,200 @@
+            new Route(Method.GET, "/internalusers/"),
+            new Route(Method.DELETE, "/internalusers/{name}"),
+            new Route(Method.PUT, "/internalusers/{name}"),
+            new Route(Method.PATCH, "/internalusers/"),
+            new Route(Method.PATCH, "/internalusers/{name}")
+        )
+    );
+
+    @Inject
+
+    public RestInternalUsersApiAction(
+        final Settings settings,
+        final Path configPath,
+        final RestController controller,
+        final Client client,
+        final ConfigurationRepository cl,
+        final ClusterService cs,
+        ThreadPool threadPool
+    ) {
+        super(settings, configPath, controller, client, cl, cs, threadPool);
+    }
+
+    @Override
+    public List<Route> routes() {
+        return routes;
+    }
+
+    @Override
+    protected Endpoint getEndpoint() {
+        return Endpoint.INTERNALUSERS;
+    }
+
+    @Override
+    protected void handlePut(RestChannel channel, final RestRequest request, final Client client, final JsonNode content)
+        throws IOException {
+
+        final String username = request.param("username");
+
+        if (username == null || username.length() == 0) {
+            badRequestResponse(channel, "No " + getResourceName() + " specified.");
+            return;
+        }
+
+        final List<String> foundRestrictedContents = RESTRICTED_FROM_USERNAME.stream()
+            .filter(username::contains)
+            .collect(Collectors.toList());
+        if (!foundRestrictedContents.isEmpty()) {
+            final String restrictedContents = foundRestrictedContents.stream().map(s -> "'" + s + "'").collect(Collectors.joining(","));
+            badRequestResponse(channel, "Username has restricted characters " + restrictedContents + " that are not permitted.");
+            return;
+        }
+
+        // TODO it might be sensible to consolidate this with the overridden method in
+        // order to minimize duplicated logic
+
+        final SecurityDynamicConfiguration<?> internalUsersConfiguration = load(getConfigName());
+
+        if (!isWriteable(channel, internalUsersConfiguration, username)) {
+            return;
+        }
+
+        final ObjectNode contentAsNode = (ObjectNode) content;
+        final JsonNode userContentAsNode = contentAsNode;
+
+        // TODO: Check role mapping and its validity here
+
+        // if password is set, it takes precedence over hash
+        final String hashedPassword = userContentAsNode.get("hash").asText();
+
+        final boolean userExisted = internalUsersConfiguration.exists(username);
+
+        // when updating an existing user password hash can be blank, which means no
+        // changes
+
+        // sanity checks, hash is mandatory for newly created users
+        if (!userExisted && hashedPassword == null) {
+            badRequestResponse(channel, "Please specify either 'hash' or 'password' when creating a new internal user.");
```

Comment:
I think this response should only say `'hash'`, based on line 122 it does not look for a field called `password`.